### PR TITLE
feat: nextjs and mdx compiler

### PR DESCRIPTION
**Pull request by ovflowd**
## Description

This PR does a significant rework on our MDX Parsing and updates Next.js to v14 and MDX to v3.

### Motivation

We used a 3rd party package to parse `mdx` with source strings, aka, `next-mdx-remote` "maintained" by Hashicorp. But as it became more apparent that we don't need a 3rd party solution and we can use a lightweight solution, I've decided to use the core of MDX itself `@mdxjs/mdx` without any extra sugar etc.

- This allows us to be always up to date with upstream `@mdxjs` and all the `unified` ecosystem (vfile, unified, remark, rehype, etc)
- Allows us to finally use `remark-gfm` plugin instead of handpicking a few versions and specific plugins and trying to make a live with it
- Removes complexity, simplifies the code, and gives better ownership for our needs.
- Simplifies some of our direct dependencies
- Removed rehype-raw and use plain Markdown. Because Markdown files should not mix HTML. **(PLEASE DOUBLE CHECK PAGES ARE WORKING AS EXPECTED (compare with nodejs.org))**
- Tried to optimise a few pieces of the build process

The implementation is bare-bones and very simple, and it works fantastically.

## Validation

Pages should be rendered as usual without content change...

### CODE REVIEW

File: next.dynamic.mjs [hunk 1]
```diff
@@ -138,41 +137,35 @@ export const generateStaticProps = async (source = '', filename = '') => {
   // by default a page is not found if there's no source or filename
   const staticProps = { notFound: true, props: {}, revalidate: false };
 
+  console.log(filename);
```
Comment (bmuenzenmeyer) [inline next.dynamic.mjs]: do you intend to leave  these console logs here?
```suggestion
```

File: next.dynamic.mjs [hunk 2]
```diff
@@ -138,41 +137,35 @@ export const generateStaticProps = async (source = '', filename = '') => {
...
+
   // We only attempt to serialize data if the `source` has content and `filename` has content
   // otherwise we return a 404 since this means that it is not a valid file or a file we should care about
   if (source.length && filename.length) {
+    console.time('createVFile');
```
Comment (bmuenzenmeyer) [inline next.dynamic.mjs]:
```suggestion
```

File: next.dynamic.mjs [hunk 3]
```diff
@@ -138,41 +137,35 @@ export const generateStaticProps = async (source = '', filename = '') => {
...
-    const frontmatter = JSON.parse(JSON.stringify(rawFrontmatter));
-
-    // this defines the basic props that should be passed back to the `DynamicPage` component
-    // We only want the `compiledSource` as we use `MDXProvider` for custom components along the journey
-    // And then we want the frontmatter and heading information from the VFile `data`
-    staticProps.props = { content: compiledSource, headings, frontmatter };
+    console.timeEnd('createVFile');
+
+    console.time('compileMDX');
+
```
Comment (bmuenzenmeyer) [inline next.dynamic.mjs]:
```suggestion
```

File: next.dynamic.mjs [hunk 4]
```diff
@@ -138,41 +137,35 @@ export const generateStaticProps = async (source = '', filename = '') => {
...
+    // This compiles our MDX source (VFile) into a final MDX-parsed VFile
+    // that then is passed as a string to the MDXProvider which will run the MDX Code
+    const { content, headings, frontmatter } = await compileMDX(
+      sourceAsVirtualFile,
+      fileExtension
+    );
+
+    console.timeEnd('compileMDX');
+
```
Comment (bmuenzenmeyer) [inline next.dynamic.mjs]:
```suggestion
```

**Review by bmuenzenmeyer — DISMISSED** [review]
js syntax highlighting is not working

✅ https://nodejs.org/en/learn/manipulating-files/working-with-folders-in-nodejs
❌ https://nodejs-org-git-feat-next-14-lightweight-mdx-compiler-openjs.vercel.app/en/learn/manipulating-files/working-with-file-descriptors-in-nodejs

at times these pages don't render. cannot reliably recreate yet. LH seemed to have the same issue

❌ https://nodejs-org-git-feat-next-14-lightweight-mdx-compiler-openjs.vercel.app/en/about

❌ https://nodejs-org-git-feat-next-14-lightweight-mdx-compiler-openjs.vercel.app/en/blog/vulnerability/openssl-fixes-in-regular-releases-oct2023

**Comment by ovflowd** [comment]
@bmuenzenmeyer feel free to re-review this ✨

**Review by AugustinMauroy — APPROVED** [review]
SGTM

**Review by canerakdas — APPROVED** [review]
LGTM, I compared the preview with nodejs.org. Also, fixes the markdown formatted tables

**Pull request merged**